Add "No Release" mod

mcendu opened the pull request:
This pull request adds a new osu!mania mod, No Release, that relaxes tail judgements. The current implementation automatically awards Perfect (or Meh if the hold note is broken midway) for a hold note tail once it crosses the hit target.


Todo:

- [x] Add tests. This should probably derive from the hold note input testsuite.

mcendu committed:
Initial implemention of the No Release mod

This commit adds a new osu!mania mod No Release that relaxes tail
judgements. The current implementation automatically awards Perfect
(or Meh if the hold note is broken midway) for a hold note tail at
the end of its Perfect window, as long as it is held by then.

Tests are pending for the next commit.

mcendu committed:
add tests for No Release mod

The new test scene is essentially a copy of TestSceneHoldNoteInput,
modified to test the judgement changes applied by the new mod. A base
class might need to be abstracted out for them.

mcendu committed:
formatting

bdach commented on osu.Game.Rulesets.Mania/Objects/Drawables/DrawableHoldNoteTail.cs:
```diff
@@ -18,6 +19,11 @@ public partial class DrawableHoldNoteTail : DrawableNote
 
         protected internal DrawableHoldNote HoldNote => (DrawableHoldNote)ParentHitObject;
 
+        /// <summary>
+        /// The minimum uncapped result for a late release.
+        /// </summary>
+        public HitResult LateReleaseResult { get; set; } = HitResult.Miss;
```
I personally have huge objections to this existing and will resist very hard to this getting anywhere near merge with this member present. It would be in the best interests of this pull to try and eliminate the need for it.

mcendu replied:
How to apply changes to hit results? Is there any existing mechanism or should I make a new mod hook?

bdach replied:
Closest thing we currently have would be the osu! mod Strict Tracking that subclasses DHOs to get the desirable result.

mcendu committed:
Rewrite no release mod

Per the request of spaceman_atlas, the No Release mod is rewritten to
avoid modifications to DrawableHoldNoteTail. The approach is based
on that of the Strict Tracking mod for the osu!(standard) ruleset,
injecting the mod behavior by replacing the normal hold note with
the mod's variant. The variant inherits most bevaior from the normal
hold note, but when creating nested hitobjects, it creates its own
hold note tail variant instead, which in turn is used to instantiate
the mod's variant of DrawableHoldNoteTail with a new behavior.

The time a judgement is awarded is changed from the end of its
Perfect window to the time of the tail itself.

smoogipoo reviewed (approved):
lgtm
